internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const lokiParserSize = 4 * 1024
+
+type lokiForwarder struct {
+	serviceName string
+	targetName  string
+	target      plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	stopCh   chan struct{}
+
+	buffer *lokiBuffer
+	client *lokiClient
+}
+
+func newLokiForwarder(
+	serviceName, targetName string, target plan.LogTarget, iterator servicelog.Iterator,
+) (*lokiForwarder, error) {
+
+	labels := map[string]string{
+		"pebble_service": serviceName,
```

Comment:
Can we make this label a constant so we can start grouping them and write docs for their purpose?

Reply:
Not going to bother, since it will (probably) be removed once we add custom label support.